drivers: led: is31fl3197 Arduino GIGA

KurtE opened the pull request:
Start of support for the is31fl3197 LED controller used by Arduino in the Arduino Giga Display shield.

So far there is just enough implemented that allows me to control the LEDS, to the same extent  as our version of the library we implemented that runs on ArduinoCore-zephyr

@pillo79 @facchinm - I am not sure what direction Arduino is heading with functionality such as this.
Add support to zephyr?  Or do it external as libraries within ArduinoCore-zephyr (either integrated within the zephyr builds or external libraries),  This is also true, about for some of the other 
features of the display, like the display itself.  Should there be samples setup to actually 
display something on the display, preferably without the need for LVGL.

So far this only has the minimal support for setting up to 4 LEDS (the GIGA only uses 3).

Added sample sketch that runs on the GIGA.  Currently it works with just
west build -p -b arduino_giga_r1//m7

I put the device tree into sample/boards.  Potentially it should be defined as part of the 
overlay in the arduino_giga_display_shield, in which case the west build should 
define the shield as part of the command.

Let me know know what you think.
Thanks

simonguinot reviewed (commented):
Hi @KurtE,

Thanks for this PR !

Can it be merged with the driver is31fl3194 ? into a is31fl319x driver ?

I didn't check but I am asking because in Linux the leds-is31fl319x driver supports the IS31FL319{0,1,3,6,9} models...

KurtE commented:
Thanks @pillo79 @mjs513,

>Thanks Kurt! I also think this is worth merging with the existing driver

Can do.  At least the basics, that the display shield uses.  Don't have any other things to try more advanced stuff on.

My first thoughts were to hack it and drive everything off of Product ID.  I believe the 94 has a fixed ID of 0xce
Whereas the 97s ID is the I2C 8 bit address: Currently checked like:
`	if (prod_id != (config->bus.addr << 1)) {
`

> Oh, and re/ Arduino attitude, upstream is the way to go so 👍 on that! By all means make it part of the Giga Display shield, that way it will be automatically available everywhere you use that and not only in the test code 🙂

I agree, that in the Zephyr world, having it all here under the shield is the way to go and not have to put the DT info in the
test sketch.   I had it there to begin with, with the testing, will move it back there.  

Some questions on this, should I then remove the GIGA from the integration section of the sample.yaml?  As I don't believe
there are any other examples for them?

Include Portenta H7? I have a overlay I experimented with yesterday, that had the LEDS working on an H7 plugged into a MID 
Carrier, which has a display shield plugged into it.  But it only worked if I put jumpers to I2C pins.  I tried it as the Display Shield
schematic that looked like there was a connection for Portenta with I2C pins... Just not sure how.
Guessing probably punt for now.

**Main Secondary Question**:

Currently with ArduinoCore-zephyr if I wish to build for a GIGA, it includes all of the Giga Shield stuff, regardless if I actually have a shield.  

With the MBED version, only the basics for the display are included with a board release in the library Arduino_H7_video, The
rest of the functionality is contained in libraries you need to install.  

I decided to try first with the LEDS as it is pretty self contained and not very large.  Our zephyr version of Arduino library for this:
```
#include "GigaDisplayRGB.h"

GigaDisplayRGB::GigaDisplayRGB(){
}

void GigaDisplayRGB::begin()
{
    Wire1.begin();
    writeByte(0x1, 0xF1);
    writeByte(0x2, 0xFF);
}

void GigaDisplayRGB::on(uint8_t r, uint8_t g, uint8_t b)
{
    writeByte(0x10, r);
    writeByte(0x11, g);
    writeByte(0x12, b);
    writeByte(0x2b, 0xc5);
}

void GigaDisplayRGB::off()
{
    on(0, 0, 0);
}

void GigaDisplayRGB::writeByte(uint8_t subAddress, uint8_t data)
{
    Wire1.beginTransmission(0x50);
    Wire1.write(subAddress);
    Wire1.write(data);
    Wire1.endTransmission();
}
```
Longer term question: how much of the stuff in the ArduinoCore-zephyr overlays/config files maybe should be
migrated upstream, but that is beyond this scope.

KurtE commented:
Quick question:

Current driver is pretty hard coded to RED/GREEN/BLUE and if I read it correctly it is assumed that the user
always has to pass in 3 colors in that order, but the output code may change the order of them as maybe the
channels to the hardware are not wired that way? 

Now with the 97 it can handle up to 4 channels, I started off hard coding the 4th channel to WHITE for maybe
RGBW type LEDS.

But the LED Header file defines other colors as well:
```
#define LED_COLOR_ID_WHITE      0
#define LED_COLOR_ID_RED        1
#define LED_COLOR_ID_GREEN      2
#define LED_COLOR_ID_BLUE       3
#define LED_COLOR_ID_AMBER      4
#define LED_COLOR_ID_VIOLET     5
#define LED_COLOR_ID_YELLOW     6
#define LED_COLOR_ID_IR         7
#define LED_COLOR_ID_MAX        8
```

There feels like there should be way to specify different colors?  And maybe mappings? 
Ok to ignore for now?

Also is there any reason things are hard coded to have to set all of the LEDS each time?  That is should it be
allowed to only set the Green LED, by passing in index 1, count 1? 
That is can there be a few different pieces of code that maybe change a state of one LED.  Sort of along the line that
GIGA has similar like LED (minus driver)
<img width="296" height="224" alt="image" src="https://github.com/user-attachments/assets/ce4584ae-1c05-4546-bf29-e428db1896c4" />
And each of the three colors are often driven by different places in the code base.

Thanks

KurtE commented:
@pillo79 @mjs513 @simonguinot  - I pushed up a version that combined it with the 94 driver.

There are probably a few things to cleanup, 

But it now builds and runs on my GIGA/display shield

simonguinot commented:
> Quick question:
> 
> Current driver is pretty hard coded to RED/GREEN/BLUE and if I read it correctly it is assumed that the user always has to pass in 3 colors in that order, but the output code may change the order of them as maybe the channels to the hardware are not wired that way?
> 
> Now with the 97 it can handle up to 4 channels, I started off hard coding the 4th channel to WHITE for maybe RGBW type LEDS.

Yes I read the code and there is a confusion about the `color-mapping` DT property and how it is used in the `is31fl3194` driver and here too.

It just describes the hardware color mapping for users. A firmware application can use it to build the `color` buffer passed to the ->set_color API method.

Here is an example of LED child node:

```
led {
        color-mapping =
                <LED_COLOR_ID_BLUE>,
                <LED_COLOR_ID_GREEN>,
                <LED_COLOR_ID_RED>;
};
```

It indicates that the first LED channel is for the blue color, the second for the green and the third for the red. And the user is responsible for building correctly the `color` buffer based on this information.

It should not be used by drivers to enforce an RGB mapping on the `color` buffer. `color[i]` goes to channel `i` and that's it.

Actually, `color-mapping` should not be used by LED drivers at all.

Hope this helps :)

simonguinot commented:
> Sorry, another quick question: Is there any Boards other than the Nicla Sense ME that uses is31fl3194? (I don't have one of those to test that I did not break it)

I don't have one. We'll try to find someone to test among the contributors of the is31fl3194 driver.

pillo79 reviewed (commented):
Thanks for the effort! Here's a first pass at the current state of the driver. Please do submit this in several commits; at a minimum:
- one to switch _the naming only_ of the current driver 3194 to 319x everywhere, **including the current test** (but no other changes!) 
- one initial refactor of the safety / RGB checks code, as per the main thread discussion 
- one (or more) that add the 3197 code,  DT changes etc, as appropriate
- one that introduces the new device in the Giga Display shield.

Note that even if CI tests the full PR, it should be possible to build Zephyr on each commit in the series (obviously with incomplete features!); otherwise it would create issues for people bisecting _other_ changes.

simonguinot commented:
> Thank you, that helps me as well, I know I had the same confusion while developing the driver. I am OK with changing / fixing the behavior of the 3194 as well, as this description makes a lot more sense than the current channel shuffling.
> 
> But then, I wonder what was the point of the `count_colors` check? IIRC that was added during the reviews of the 3194 driver. We should drop it as well, right?

It was maybe to differentiate the LED mode handling: three monochromatic LEDs against a single multicolor LED mode ?

KurtE commented on drivers/led/is31fl3194.c:
```diff
@@ -78,23 +132,24 @@ static int is31fl3194_get_info(const struct device *dev,
 
-	if (num_colors != 3) {
-		return -EINVAL;
+	if (info->num_colors > config->channel_count) {
```
done - will take care of some of this low hanging fruit before I split up the changes

KurtE commented on drivers/led/is31fl3194.c:
```diff
@@ -22,6 +21,23 @@
 LOG_MODULE_REGISTER(is31fl3194, CONFIG_LED_LOG_LEVEL);
+#define REG_NOT_DEFINED 0xff
+
+struct is31f1319x_regs {
```
done

KurtE commented:
@pillo79 - I pushed up a lot of the changes you requested.  Still all combined.
Did not change color model yet.  

Next up to undo everything and start logically over in pieces.

KurtE commented:
> Thanks for the effort! Here's a first pass at the current state of the driver. Please do submit this in several commits; at a minimum:
> 
> * one to switch _the naming only_ of the current driver 3194 to 319x everywhere, **including the current test** (but no other changes!)
> * one initial refactor of the safety / RGB checks code, as per the main thread discussion
> * one (or more) that add the 3197 code,  DT changes etc, as appropriate
> * one that introduces the new device in the Giga Display shield.

Sorry wondering if it needs to be in the above order?  Might it  also be done:
> 
> * one to switch _the naming only_ of the current driver 3194 to 319x everywhere, **including the current test** (but no other changes!)
> * one (or more) that add the 3197 code,  DT changes etc, as appropriate
> * one that introduces the new device in the Giga Display shield.
> * one initial refactor of the safety / RGB checks code, as per the main thread discussion

As I am still unsure of how the mapping should work.

As the Nicla Sense ME has the LEDS like: BGR
<img width="403" height="179" alt="image" src="https://github.com/user-attachments/assets/33054fcc-bbb3-47f2-95aa-d14c8daab36d" />
And the Giga is like: RGB
<img width="431" height="356" alt="image" src="https://github.com/user-attachments/assets/f9f59b41-129d-41fd-ac5e-826982b4ca0d" />

But I would hope that the application code should not have to know that implementation detail

KurtE committed:
drivers: led: rename IS31FL3194 to 9x

In preparation to allow this driver to support more than
one IS31FL319x object such as ...97, rename the
source file and the functions to make them more
generic.

I  also changed the enumeration/creation of object
to generate the right names for the current 94 objects

Renamed the Kconfig file. Updated CMakeLists.txt to
check for different condition to build source file

The samples driver is31fl3194 was also renamed to
is31fl319x. All of the references to use 9x instead
of 94. The updated sample required an update to the
doc script redirect.py to reflect the new name.

Signed-off-by: Kurt Eckhardt <kurte@rockisland.com>

KurtE commented:
Sorry did not mean to close it

KurtE committed:
drivers: led: is31fl319x - hard coded registers

more refactoring, don't use the 94 registers everywhere.
Introduce structure, with registers and fixed values, and use it.

Signed-off-by: Kurt Eckhardt <kurte@rockisland.com>

KurtE committed:
drivers: led: is31fl319x - redo color to channel

Removed the driver knowing anything about what color each channel is.
When you output to an LED with a buffer it simply writes the
channels out with buffer.

Also removed the assumptions that the leds are either
a) RGB  or b) 3 channels and instead use them the way they are defined.

That is for example if we have 4 channels, we could define an LED
with just one color in it and we could define another with three.

The user can define 4 leds each with color yellow.

For each LED it counts how many buffers were used by previous LEDs and
starts there and outputs N channels.

Decided it looked cleaner to use helper function write channels, so
implemented it and added it to the call table.

Updated sample:
It manually computes which channel maps to red, green and blue and then
uses those indexes to map the table of colors to the right channels.
Signed-off-by: Kurt Eckhardt <kurte@rockisland.com>

KurtE commented:
@pillo79 @mjs513 @simonguinot - I split this up into multiple commits.
I also redid the color mapping... i.e. there is none.

I have the GIGA display shield working with it.
The Example: I updated to look through the channels, to figure out which was the R and G and B,
And then the sample remaps the colors.

KurtE committed:
boards: shields: arduino_giga_display_shield: add LED to DT

Added is31fl3197@50 to overlay

Also updated GIGA board yaml file to say that it supports i2c

And updated sample to mention it

Signed-off-by: Kurt Eckhardt <kurte@rockisland.com>

KurtE commented:
> Please retry analysis of this Pull-Request directly on SonarQube Cloud

how?

pillo79 reviewed (commented):
Quite a better shape, thanks for the effort! :+1: 

Note that commit ' drivers: led: is31fl319x - redo color to channel ' introduces changes to the sample `README.rst`  that are later removed and then re-added in the following commits. :eyes:  Ideally, at that stage there should be no mention of other chips yet.

KurtE commented on drivers/led/CMakeLists.txt:
```diff
@@ -6,7 +6,7 @@ zephyr_library()
 # zephyr-keep-sorted-start
 zephyr_library_sources_ifdef(CONFIG_HT16K33 ht16k33.c)
-zephyr_library_sources_ifdef(CONFIG_IS31FL3194 is31fl3194.c)
+zephyr_library_sources_ifdef(CONFIG_IS31FL3194 is31fl319x.c)
```
Thanks, 

When I was refactoring, I tried renaming the CONFIG file as mentioned, but at that time it would not build,
I got some strange error about missing some node or the like, when I did not rename the config file it worked.
Maybe something else was going wrong. 

I will try again.

pillo79 replied:
> The problem is that the defines like: CONFIG_IS31FL3194 and CONFIG_IS31FL3197 are not longer defined.

`DEVICE_DT_GET_ANY` will happily return NULL if it doesn't find any device with the specified compatible. You could have vars defined for both devices, and then pick the one that is actually not-NULL :wink:

KurtE replied:
I resolved it... Had to change the CMakeLists to look for CONFIG_IS31FL319X instead of 4 or 7

Changed to check for null in app.

KurtE commented on drivers/led/is31fl319x.c:
```diff
@@ -250,20 +288,30 @@ static int is31fl319x_init(const struct device *dev)
-		LOG_ERR("%s: invalid product ID 0x%02x (expected 0x%02x)", dev->name, prod_id,
-			regs->prod_id_val);
-		return -ENODEV;
+	if (regs->prod_id_val != 0xff) {
```
Can do,  I sort of resisted that earlier as have not looked through other chips in same family to know if these
differences are unique or similar for others as well.  The other way I thought about doing it was:
add new item to config structure, like: uint8_t features and have either defines or enum with bit flags.
Like: 
#define FEATURE_ID_IS_ADDR 0x01
#define FEATURE_SET_CURRENT 0x02

And have the enumerator at end for each chip pass in the appropriate flags.
Worth doing?  Or wait until (if) other chips are supported?

KurtE replied:
done

KurtE commented on drivers/led/Kconfig.is31fl3197:
done

KurtE commented on samples/drivers/led/is31fl319x/README.rst:
```diff
@@ -25,22 +25,15 @@ Building and Running
 any board where the devicetree has an I2C device node with compatible
-:dtcompatible:`issi,is31fl3194` enabled, along with the relevant bus
-controller node also being enabled.
+:dtcompatible:`issi,is31fl3194` or :dtcompatible:`issi,is31fl319` with the
```
done - Wish I know better ways to resolve simple things like this.  I had fixed it in the next commit, but...
So made an edit and then when through a squash of both places and edited the conflicts

thedjnK commented on drivers/led/is31fl319x.c:
```diff
@@ -319,25 +318,28 @@ static DEVICE_API(led, is31fl3194_led_api) = {
 										\
-	DT_INST_FOREACH_CHILD(id, COLOR_MAPPING)				\
+	static const struct led_info is31fl319##id##_leds_##n[] =		\
+		{ DT_INST_FOREACH_CHILD(n, LED_INFO) };			\
```
issue here seems to be you are fixing in wrong commit, check first commit here, the line ending on lines 321 and 326 are wrong, they are wrongly added in the first commit then fixed in the second commit, can you fix them in the first commit?

thedjnK reviewed (commented):
PR needs original commit fixing instead of fixups in second commit then looks OK

thedjnK commented:
> > PR needs original commit fixing instead of fixups in second commit then looks OK
> > The other option might be to squash those two commits into one?
> 
> Edit: Is this about the \ don't align?

Look at first 2 commits, anything that the first commit adds and the second commit fixes up, move to the first commit

thedjnK commented:
First commit, line 343 in .c file, undef it still present

JarmouniA reviewed (changes requested):
First & 3rd commit message needs fixing, there is stuff after the sign-off

KurtE commented:
> First & 3rd commit message needs fixing, there is stuff after the sign-off
I think done now ... Had git hiccup ... Github desktop decided to do a merged... I think
that was undone now...

pillo79 reviewed (dismissed):
Thanks!

simonguinot commented on drivers/led/Kconfig.is31fl319x:
```diff
@@ -1,8 +1,8 @@
 # Copyright (c) 2024 Arduino SA
```
Can you please fix the commit message ? It claims that the Kconfig file was not renamed, while it is. And also there is a lot of garbage I don't understand...

KurtE replied:
Updated first commit message

simonguinot commented on samples/drivers/led/is31fl319x/src/main.c:
```diff
@@ -32,20 +33,42 @@ static const struct color_data color_sequence[] = {
-		if (ret < 0) {
+		/* Map the colors */
+		if (!output_rgb_color(led_dev, 0, &(color_sequence[i].r))) {
+			printk("Call to led_set_color failed\n");
```
I don't understand how this whole sample works. It assumes everywhere that the LEDs are RGB. But it is not necessarily the case.

Since we are claiming that the driver supports both monochromatic and multi-color LEDs:

```
That is for example if we have 4 channels, we
could define an LED with just one color in it
and we could define another with three.
```

Then it should be supported by the sample as well.

KurtE replied:
@simonguinot (and all), I totally understand what you are saying, about it not showing specific sample for non-RGB setup of LEDS.  

But currently the only two boards (both Arduino) that I know of use them as RGB drivers.  And they are wired up differently (one sort or RGB and the other BGR), but that is completely arbitrary.  They GIGA could have been wired as, G, unused, R, B, if that turned out to be more convenient. 

The original example simply cycled through a set of colors, like this, the main parts I added was to fade first Red, then Green, then Blue led so you can find out if your coding of the LED colors matches the hardware... Which is how I found
one as BGR and other RGB.

Was not sure what to add?  As arbitrary hardware could do any number of things, like groupings of:
1111(no grouping), 211, 121, 112, 22, 31, 13 if that makes sense.   So feel like maybe additional or replacement 
tests could be added for different configurations as they come along.   In the same way as some cameras 
support output of JPEG, but potentially an JPEG example might be added as when a hardware setup is 
created that makes use of that capability.

I also wonder how much this current abstraction makes sense to use for something like the Arduino Giga Display Shield.
That is the current code, (which I did not define) I use in the Arduino world is pretty simple:
```
class GigaDisplayRGB {
    public:
        GigaDisplayRGB();
        void begin();
        void on(uint8_t,uint8_t,uint8_t);
        void off();
    private:
        void writeByte(uint8_t,uint8_t);
};
```
and
```
#include "GigaDisplayRGB.h"

GigaDisplayRGB::GigaDisplayRGB(){
}

void GigaDisplayRGB::begin()
{
    Wire1.begin();
    writeByte(0x1, 0xF1);
    writeByte(0x2, 0xFF);
}

void GigaDisplayRGB::on(uint8_t r, uint8_t g, uint8_t b)
{
    writeByte(0x10, r);
    writeByte(0x11, g);
    writeByte(0x12, b);
    writeByte(0x2b, 0xc5);
}

void GigaDisplayRGB::off()
{
    on(0, 0, 0);
}

void GigaDisplayRGB::writeByte(uint8_t subAddress, uint8_t data)
{
    Wire1.beginTransmission(0x50);
    Wire1.write(subAddress);
    Wire1.write(data);
    Wire1.endTransmission();
}
```
Which could improved,  The replacement code using this hardware abstraction, would require parts of this example sketch to map the R, G, B colors to their proper index to then output using the zephyr functions. 

So currently this is my lowest priority.  Currently being interrupted by a very high priority and demanding task.
a Puppy!

simonguinot replied:
One purpose of the sample is to demonstrate the driver. Here is a suggestion:

In a loop, cycle through the different LEDs. And for each LED, use the `->get_info` API method to retrieve and print information about the LED. If it is a monochromatic LED, turn it on and off, using `->set_colors()`. And `with ->set_brightness` turn it on again, gradually to its maximum level, If it is multi-color LED, do the same, but for each color. And as a last test, apply the very same pattern, but to all channels simultaneously, using `->write_channels()`.

This would allow to cover the different DT configurations (monochromatic, multi-color, mixed) and test all the LED API methods implemented by the driver.

simonguinot commented on drivers/led/is31fl319x.c:
```diff
@@ -161,11 +152,32 @@ static int is31fl319x_set_color(const struct device *dev, uint32_t led, uint8_t
+		return -ENODEV;
+	}
+
+	channel_start = is31fl319x_map_led_to_start_channel(config, led);
```
I think there is a confusion here. The ->set_color() API method allows to configure the colors of a LED (for a given index). It should not overflow on the next LED. You can't handle LEDs as channels here.

simonguinot replied:
OK, understood.

Please use the following test instead:

```
        if (num_colors != info->num_colors) {
                LOG_ERR("%s: invalid number of colors: got=%d, expected=%d",
                        dev->name,
                        num_colors,
                        info->num_colors);
                return -EINVAL;
        }
```

The user must configure exactly the same number of colors that supported by the LED. If not, then `-EINVAL` is the correct error code to return.

In addition, please move this check before the call to `is31fl319x_map_led_to_start_channel().`

simonguinot reviewed (changes requested):
Hi @KurtE,

I apologize for the delay and also for not following carefully this PR for a while.

Please find my review below. It is only a start.

In addition, could you try to fix the commit splitting and improve the message wording ? It's quite disorganized and that doesn't help with the review.

KurtE commented on samples/drivers/led/is31fl319x/src/main.c:
```diff
@@ -51,7 +52,14 @@ int main(void)
+	}
+
 	if (!device_is_ready(led_dev)) {
+		printk("No devices were found");
```
Sorry, I am not sure what is wanted here now.  We have:
```
const struct device *led4_dev = DEVICE_DT_GET_ANY(issi_is31fl3194);
const struct device *led7_dev = DEVICE_DT_GET_ANY(issi_is31fl3197);
,,,
	const struct device *led_dev = led4_dev;

	if (led_dev == NULL) {
		led_dev = led7_dev;
	}

	if (!device_is_ready(led_dev)) {
		printk("No devices were found");
		return 0;
	}
```
So if the board the user builds for has an is31fl3194 defined for it, the led_dev should be set to the first one
if not it is NULL and it assigns to led_dev the first ...97 led if Any.

At one point I would have tested that for NULL and errored, but in previous reviews was told to remove the
redundant check as device_is_ready will check for NULL.    So are you now asking for me to add the 
check for NULL back in?   

Note: this sample, like I believe most will error out if the device is marked as delay init. As I don't then
try to call the device initialization function

simonguinot replied:
> At one point I would have tested that for NULL and errored, but in previous reviews was told to remove the redundant check as device_is_ready will check for NULL. So are you now asking for me to add the check for NULL back in?

I suggest something like this:
```
        if (!led_dev) {
                printk("No device with compatible issi,is31fl3194 or issi,is31fl3197 found");
                return 0;
        }
        if (!device_is_ready(led_dev)) {
                printk("LED device %s not ready", led_dev->name);
                return 0;
        }
```
This provides the user with more information about the error.

As I said I haven't follow all the comments, and I already apologized for that. There were a lot of them. And I am sorry for missing this one.

KurtE replied:
Pushed up changes, with slight mods.
Split the first printk into multiple printks as I think it caused the line to be too long.

Also added \n's to them

simonguinot commented on samples/drivers/led/is31fl319x/src/main.c:
```diff
@@ -32,20 +33,42 @@ static const struct color_data color_sequence[] = {
+/*
+ * Main
+ */
 int main(void)
```
> Moved, in next push.
> 
> Fascinating... Most people over the decades have wanted the most important thing first and the implementation details later... But...

Hi @KurtE,

Given the number of comments and previous approvals, I expected this PR to be in much better shape. I started my review fairly confident that I would only find "minor" issues. That's why minor comments are mixed with more serious concerns…

And please, don't resolve conversations. It is up to the person who posted a comment to close it, if he is happy with the resolution.

thedjnK commented on drivers/led/is31fl319x.c:
```diff
@@ -280,9 +233,10 @@ static int is31fl319x_init(const struct device *dev)
 	const struct is31f1319x_model *regs = config->regs;
-	int i, ret;
-	uint8_t prod_id, band;
+	int i, j, ret;
```
one variable per line as per coding guidelines

thedjnK commented on drivers/led/is31fl319x.c:
```diff
@@ -323,12 +392,21 @@ static DEVICE_API(led, is31fl319x_led_api) = {
+#endif
+
+#ifdef CONFIG_DT_HAS_ISSI_IS31FL3197_ENABLED
+#undef DT_DRV_COMPAT
```
then move this undef after also, makes it neater

KurtE commented on drivers/led/Kconfig.is31fl319x:
```diff
@@ -1,8 +1,8 @@
-config IS31FL3194
-	bool "IS31FL3194 LED driver"
+config IS31FL319X
+	bool "IS31FL319X LED driver"
```
done will be in next push

simonguinot reviewed (commented):
Hi @KurtE 

The driver itself is in a pretty good shape now. Well done.

I only have minor comments. The main thing is the sample. It seems really complicated to me and I suggest we drop the RGB mixing color test.

See my comments below.

KurtE commented on samples/drivers/led/is31fl319x/src/main.c:
```diff
@@ -151,3 +202,108 @@ bool output_rgb_color(const struct device *dev,  uint32_t led, const uint8_t *co
+	fade_all_channels(led_dev, device_channel_count);
+
+	/* Check for RGB LED */
+	rgb_led = generate_rgb_color_mapping(led_dev);
```
I am sort of torn here.  That is the reason that code is here is or was, one primary usage for this library was for handling RGB leds and maybe by extension RGBW.  In fact the current example sketch simply does the mapping of colors to the LEDS...

The earlier(current) library, was setup in cases where it was doing similar code to what is now in the example sketch.  That is, it would try to find if the current LED had 3 colors and locate the index for the Red, and Green and Blue leds.
Then when the user logically wanted to output an RGB output, the code would then remap lets say the first byte to the RED index, the second byte to the green index and the third byte to the Blue index.  And in the case of the two Arduino boards, one of the boards has the leds wired up like BGR and the new one board has it wired as RGB order.  But in theory they could be in any such order.

So we sort of figured that should have the code in the example that lets others know at least one way on how to do it.
Note On Arduino Side, assuming this ever goes through, not sure for the GIGA, if it makes sense to go through
these objects or simply do what it is currently doing, which is all hard coded:
```
GigaDisplayRGB::GigaDisplayRGB(){
}

void GigaDisplayRGB::begin()
{
    Wire1.begin();
    writeByte(0x1, 0xF1);
    writeByte(0x2, 0xFF);
}

void GigaDisplayRGB::on(uint8_t r, uint8_t g, uint8_t b)
{
    writeByte(0x10, r);
    writeByte(0x11, g);
    writeByte(0x12, b);
    writeByte(0x2b, 0xc5);
}

void GigaDisplayRGB::off()
{
    on(0, 0, 0);
}

void GigaDisplayRGB::writeByte(uint8_t subAddress, uint8_t data)
{
    Wire1.beginTransmission(0x50);
    Wire1.write(subAddress);
    Wire1.write(data);
    Wire1.endTransmission();
}
```
Let me know if you wish for me to remove the color tables code.  At this point I just want it done

fabiobaltieri reviewed (commented):
hey one more note, on the commit message you have 72 columns, use them? you have some paragraphs on 72, some much shorter and it makes it very hard to read

KurtE commented:
> Hi @pillo79. Please, can you test on your hardware and confirm it is working correctly ? And feel free to review and approve (again) this PR :)

It has been a couple of git rebases since I tried it on
`west build -p -b arduino_nicla_sense_me`

Looks like there are some newer independent issues with NRF...
```
nrf_rtc_timer.c
D:/Users/kurte/zephyrproject/zephyr/drivers/timer/nrf_rtc_timer.c: In function 'int_event_disable_rtc':
D:/Users/kurte/zephyrproject/zephyr/drivers/timer/nrf_rtc_timer.c:739:25: error: 'NRF_RTC_INT_COMPARE_0_MASK' undeclared (first use in this function); did you mean 'NRF_RTC_INT_COMPARE0_MASK'?
  739 |                         NRF_RTC_INT_COMPARE_0_MASK |
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                         NRF_RTC_INT_COMPARE0_MASK
D:/Users/kurte/zephyrproject/zephyr/drivers/timer/nrf_rtc_timer.c:739:25: note: each undeclared identifier is reported only once for each function it appears in
D:/Users/kurte/zephyrproject/zephyr/drivers/timer/nrf_rtc_timer.c:740:25: error: 'NRF_RTC_INT_COMPARE_1_MASK' undeclared (first use in this function); did you mean 'NRF_RTC_INT_COMPARE1_MASK'?
  740 |                         NRF_RTC_INT_COMPARE_1_MASK |
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                         NRF_RTC_INT_COMPARE1_MASK
D:/Users/kurte/zephyrproject/zephyr/drivers/timer/nrf_rtc_timer.c:741:25: error: 'NRF_RTC_INT_COMPARE_2_MASK' undeclared (first use in this function); did you mean 'NRF_RTC_INT_COMPARE2_MASK'?
  741 |                         NRF_RTC_INT_COMPARE_2_MASK |
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                         NRF_RTC_INT_COMPARE2_MASK
D:/Users/kurte/zephyrproject/zephyr/drivers/timer/nrf_rtc_timer.c:742:25: error: 'NRF_RTC_INT_COMPARE_3_MASK' undeclared (first use in this function); did you mean 'NRF_RTC_INT_COMPARE3_MASK'?
  742 |                         NRF_RTC_INT_COMPARE_3_MASK;
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                         NRF_RTC_INT_COMPARE3_MASK
[122/172] Building C object zephyr/kernel/CMakeFiles/kernel.dir/fatal.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: 'C:\Program Files\CMake\bin\cmake.EXE' --build 'D:\Users\kurte\zephyrproject\zephyr\samples\drivers\led\is31fl319x\build'
```

pillo79 reviewed (approved):
Looks great to me! @KurtE @simonguinot thank you both for this effort! 👍

simonguinot reviewed (approved):
Thank you for your patience and perseverance, @KurtE :)
